pcdsdevices/tpr.py

```diff
@@ -0,0 +1,121 @@
+from ophyd import Component as Cpt
+from ophyd import Device, EpicsSignal, EpicsSignalRO
+from ophyd import FormattedComponent as FCpt
+
+from .interface import BaseInterface
+from .pv_positioner import PVPositionerIsClose
+from .signal import MultiDerivedSignal, MultiDerivedSignalRO
+from .type_hints import SignalToValue
+
+TPR_TICK_NS = 5.384
+TPR_TAP_NS = 0.08
+
+
+def _get_delay(mds: MultiDerivedSignal, items: SignalToValue) -> float:
+    """Calculates delay in ns from ticks and taps"""
+    return items[mds.attrs[0]] * TPR_TICK_NS + items[mds.attrs[1] * TPR_TAP_NS]
+
+
+def _get_width(mds: MultiDerivedSignal, items: SignalToValue) -> float:
+    """Calculates width in ns from ticks"""
+    return items[mds.attrs[0]] * TPR_TICK_NS
+
+
+def _put_last(mds: MultiDerivedSignal, value: float) -> SignalToValue:
+    """Only writes value to last attr"""
+    return {mds.attrs[-1]: value}
+
+
+class TprMotor(PVPositionerIsClose):
+    """
+    PV Positioner for adjusting an TPR channel.
+
+    Moves that are less than one tick
+    are considered immediately complete.
+    """
+    setpoint = FCpt(EpicsSignal, "{self.prefix}{self.sys}TDES", kind="normal")
+    delay_ticks = FCpt(EpicsSignal, '{self.prefix}{self.trg}TDESTICKS', kind="omitted")
+    delay_taps = FCpt(EpicsSignal, '{self.prefix}{self.trg}TDESTAPS', kind="omitted")
+    readback = Cpt(
+        MultiDerivedSignalRO,
+        attrs=['delay_ticks', 'delay_taps'],
+        calculate_on_get=_get_delay
+    )
+    atol = TPR_TICK_NS
+    rtol = 0
+
+    def __init__(self, prefix, *, sys, trg, name, **kwargs):
+        self.prefix = prefix
+        self.sys = sys
+        self.trg = trg
+        super().__init__(prefix, name=name, **kwargs)
+
+
+class TprTrigger(BaseInterface, Device):
+    """Class for an individual TprTrigger."""
+    ratemode = FCpt(EpicsSignal, '{self.prefix}{self.ch}RATECODE', kind="config")
+    group = FCpt(EpicsSignal, '{self.prefix}{self.ch}GROUP', kind="omitted")
+    seqcode = FCpt(EpicsSignal, '{self.prefix}{self.ch}SEQCODE', kind="omitted")
+    fixedrate = FCpt(EpicsSignal, '{self.prefix}{self.ch}FIXEDRATE', kind="omitted")
+    count = FCpt(EpicsSignal, '{self.prefix}{self.ch}COUNT', kind="omitted")
+    destmask = FCpt(EpicsSignal, '{self.prefix}{self.ch}DESTMASK', kind="omitted")
```

Comment:
Though the component names are relatively readable/understandable, would you mind going through these and explaining the purpose of each through `Component(..., doc="description goes here")` keyword arguments?